replica/table.cc

```diff
@@ -3176,8 +3176,15 @@ void table::set_schema(schema_ptr s) {
     _schema = std::move(s);
 
     for (auto&& v : _views) {
-        v->view_info()->set_base_info(
-            v->view_info()->make_base_dependent_view_info(*_schema));
+        auto base_info = v->view_info()->make_base_dependent_view_info(*_schema);
```

Comment:
If what I wrote makes sense, I would add it to the commit description for posterity.

Reply:
> I think the purpose of the old code was not to update views because base table can change without the views. migration_manager issues view update mutations together view base table altering mutations.

But we did update view base infos even before this change. This patch only changes it so the updates are preserved in case the `schema_ptr` in the registry is lost and recreated.

> The purpose was atomicity of the change, so that if a view-updating operation obtains a base schema and view's base info without deferring point in between, the base schema will match the version in base_info. It is necessary, because when we start a base table read for view updates, we have one reader for all views. So we cannot take base_info's base schema_ptr and read using that. Mutations produced by that read are interpreted for each view using that view's base_info. That base_info must correspond to the base version used by the reader.
> 
> We have a check in `make_view_update_builder()` which verifies this (`base` schema_ptr comes from the mutation reader):
> 
> ```c++
>     auto vs = views_to_update | std::views::transform([&] (view_and_base v) {
>         if (base->version() != v.base->base_schema()->version()) {
>             on_internal_error(vlogger, format("Schema version used for view updates ({}) does not match the current"
>                                               " base schema version of the view ({}) for view {}.{} of {}.{}",
>                 base->version(), v.base->base_schema()->version(), v.view->ks_name(), v.view->cf_name(), base->ks_name(), base->cf_name()));
>         }
> ```
> 
> And notice how `view_update` ensures that reader's schema is upgraded to the current base table schema when view_ptr:s are collected:
> 
> ```
>             auto base_schema = _step.base->schema();
>             auto views = with_base_info_snapshot(_views_to_build);
>             auto reader = make_mutation_reader_from_fragments(_step.reader.schema(), _builder._permit, std::move(_fragments));
>             auto close_reader = defer([&reader] { reader.close().get(); });
>             reader.upgrade_schema(base_schema);
> ```
> 
> _step.base is pointer to the `table` object. `reader.upgrade_schema()` ensures that reader will generate mutations conforming to `base_schema` we just obtained, atomically with base_infos in `with_base_info_snapshot()`.

If we upgrade schema of the reader to the most recent schema in `table`, it seems like we have to have up-to-date `base_info`s for all versions of all views referring to this base table as well to be able to generate updates. I don't think even the solution mentioned at the end with updating base and view schemas atomically would help with that.

Note that this is also the code for view building, not regular view update generation. I don't see that we upgrade schema for regular view updates, and we probably don't need to - no new views relying on an up-to-date base schema can appear during regular view updates.

Aside from updating all `base_info`s the only other solution for this is losing the optimization of a common reader for all views and building views one-by-one. I actually thought this was what we were already doing (@nyh?), but it looks like we only build one step for one view at a time, but we do this step for all views before proceeding to the next one. If I'm not mistaken, the new view building coordinator will build one view at a time (at least in the initial implementation), so we'll avoid this problem.

> Changing base_info of existing view_ptr shouldn't be a problem for that code, since it obtains a snapshot of base_info, doesn't reach it via view_ptr.

From what I'm seeing, we snapshot base infos from `view_ptr`s:
```C++
std::vector<db::view::view_and_base> with_base_info_snapshot(std::vector<view_ptr> vs) {
    return vs | std::views::transform([] (const view_ptr& v) {
        return db::view::view_and_base{v, v->view_info()->base_info()};
    }) | std::ranges::to<std::vector>();
}
```
So a problematic scenario that I'm seeing is:
1) Collect `view_ptr`s
2) Update a view - this will replace the `view_ptr` in the views list in the base `table`
3) Update base - this will not update the base info in old `view_ptr`
4) When we upgrade reader schema to current base table schema, the we'll use the wrong base info from the old `view_ptr`
Though I'm not sure when we collect `view_ptr`s for the build step, maybe they're updated 

Actually I think I see how https://github.com/scylladb/scylladb/pull/21862 might have failed:
* When creating or dropping an index we seem to change the base schema
* The test failed after we dropped 2 indexes during view building
So the scenario could have been as follows:
1) We drop first index, base schema is updated and base info for all indexes, first index `view_ptr` is removed from base `table` 
2) We drop second index, base schema is updated and base info is updated only for second index
3) View building tries to finish a step, but the base info for the first index was outdated

This

> But this is not very safe, it works only when working with local schemas. re-obtaining the base_info via view_ptr will give a different base_info. It would be better to preserve snapshot semantics, it would make the code easier to reason about.
> 
> To achieve that we would have to atomically change base schema_ptr and view's schema_ptr when processing schema changes. And keep old versions of view_ptr intact (do not override their base_info).
> 
> This way we also don't have to chase view_ptr version in schema_registry.

Right, this seems like the best solution. I think I'll try implementing this somewhat soon

> But this patch works with the current approach, so looks good.

Reply:
@tgrabiec regarding 
> To achieve that we would have to atomically change base schema_ptr and view's schema_ptr when processing schema changes. And keep old versions of view_ptr intact (do not override their base_info).

Is it actually compatible with our mutation reader code?

> And notice how view_update ensures that reader's schema is upgraded to the current base table schema when view_ptr:s are collected:
```
            auto base_schema = _step.base->schema();
            auto views = with_base_info_snapshot(_views_to_build);
            auto reader = make_mutation_reader_from_fragments(_step.reader.schema(), _builder._permit, std::move(_fragments));
            auto close_reader = defer([&reader] { reader.close().get(); });
            reader.upgrade_schema(base_schema);
```
> _step.base is pointer to the table object. reader.upgrade_schema() ensures that reader will generate mutations conforming to base_schema we just obtained, atomically with base_infos in with_base_info_snapshot().

Because `view_ptr`s are collected in advance and we upgrade the readers schema for every build step, we need to update all versions of `view_ptr`s with new `base_info`s to avoid an issue here (or perhaps we could collect `view_ptr`s for every build step instead). It doesn't look like keeping old `view_ptr`s unchanged and consistent with old `base_info`s helps with anything considering we're updating the reader with the base_schema independently

Reply:
> @tgrabiec regarding
> 
> > To achieve that we would have to atomically change base schema_ptr and view's schema_ptr when processing schema changes. And keep old versions of view_ptr intact (do not override their base_info).
> 
> Is it actually compatible with our mutation reader code?

Well, with_base_info_snapshot() would have to re-obtain view ptrs, since base info inside existing view_ptr would not change. 

> 
> > And notice how view_update ensures that reader's schema is upgraded to the current base table schema when view_ptr:s are collected:
> 
> ```
>             auto base_schema = _step.base->schema();
>             auto views = with_base_info_snapshot(_views_to_build);
>             auto reader = make_mutation_reader_from_fragments(_step.reader.schema(), _builder._permit, std::move(_fragments));
>             auto close_reader = defer([&reader] { reader.close().get(); });
>             reader.upgrade_schema(base_schema);
> ```
> 
> > _step.base is pointer to the table object. reader.upgrade_schema() ensures that reader will generate mutations conforming to base_schema we just obtained, atomically with base_infos in with_base_info_snapshot().
> 
> Because `view_ptr`s are collected in advance and we upgrade the readers schema for every build step, we need to update all versions of `view_ptr`s with new `base_info`s to avoid an issue here (or perhaps we could collect `view_ptr`s for every build step instead). It doesn't look like keeping old `view_ptr`s unchanged and consistent with old `base_info`s helps with anything considering we're updating the reader with the base_schema independently

It doesn't help now, but simplifies reasoning and maybe avoids issues in the future.